### PR TITLE
Add known_publish_date_tags to goose configuration

### PR DESCRIPTION
**Pull request by dlrobertson**
## Summary

 - Loosely enforce PEP 8
 - Use a class instead of a `dict` for the article context and publish date patterns
 - Move the static `KNOWN_PUBLISH_DATE_TAGS` list to the `Configuration` class
with a property and setter function allowing users to configure custom
tags for lesser used publish_date tags.

## Example

```python
import warnings
from goose3 import Goose, Configuration, PublishDatePattern

warnings.simplefilter('always', DeprecationWarning)

config = Configuration()

config.known_publish_date_tags = {'attribute': 'name', 'value': 'pubdate', 'content': 'content',
                                  'domain': 'sputniknews.com'}
# or
config.known_publish_date_tags = PublishDatePattern(attr='name', value='pubdate', content='content',
                                                    domain='sputniknews.com')

g = Goose(config)
article = g.extract(url="https://sputniknews.com/middleeast/201807281066762466-syrian-kurds-government-agreement/")

print(article.publish_date)                                                                                                                                                                                                           
```

### CODE REVIEW

**Comment by dlrobertson** [comment]
Follow-up from https://github.com/goose3/goose3/pull/55#issuecomment-407404766

**Comment by dlrobertson** [comment]
@barrust @lababidi any thoughts on this? I expanded the initial scope of this PR a bit.

**Comment by dlrobertson** [comment]
>  I am concerned about changing the behavior of how we allow new context managers.

That is definitely valid.

> Not sure if you throw the deprecation warning on that or not, but it is something to think about.

Yeah, I handle the case that the old method is used. The implementation here should be backwards compatible. I'll add tests for this.

On second thought perhaps there should be no deprecation warning and we should explicitly allow users to either add a `dict` or the new class for the pattern.

**Comment by barrust** [comment]
> On second thought perhaps there should be no deprecation warning and we should explicitly allow users to either add a dict or the new class for the pattern.

I think that would be best, allow for both methods: dict or class. Thanks!

File: goose3/configuration.py [hunk 1]
```diff
@@ -57,7 +95,8 @@ def __init__(self):
 
         # extraction information
         self._local_storage_path = os.path.join(tempfile.gettempdir(), 'goose')
-        self._known_context_patterns = KNOWN_ARTICLE_CONTENT_PATTERNS
+        self._known_context_patterns = KNOWN_ARTICLE_CONTENT_PATTERNS[:]
```
Comment (dlrobertson) [inline goose3/configuration.py]: Fixes bug. We need to ensure we do a deepcopy here instead of just copying the mutable pointer.

**Comment by dlrobertson** [comment]
If it makes it easier to review/merge I can remove the feature that allows for filtering a pattern for a domain.

**Comment by barrust** [comment]
No, it is more just a time thing. I will get to it by the end of the weekend. I have done a cursory glance at the code and didn't see anything out of the ordinary. 

I am not sure there is a need for the tox.ini file as we already have a setup.cfg file that can hold the flake8 configuration information!

**Comment by dlrobertson** [comment]
> No, it is more just a time thing. I will get to it by the end of the weekend.

Np. I totally understand!

> I am not sure there is a need for the tox.ini file as we already have a setup.cfg file that can hold the flake8 configuration information!

Moved the `[flake8]` section to `setup.cfg`.

**Pull request merged**